### PR TITLE
Fix printing of `AbstractDict`s with unknown length

### LINKED ISSUE #55931: `Base.show(::IO, ::MIME"text/plain", d::AbstractDict)` requires `length(d)`
`Base.show(::IO, ::MIME"text/plain", d::AbstractDict)` for which `length(d)` is not defined crashes.

It crashes even if `Base.IteratorSize(typeof(d)) isa Base.SizeUnknown`, which is the standard way for an iterator to indicate that it doesn't support `length`.

It seems like `AbstractDict` doesn't require `length` to be defined in general.  For example, `Base.summary(::IO, ::AbstractDict)` handles the case where the dictionary doesn't have `length`.

Reproduction:
```
struct NoLengthDict{K,V} <: AbstractDict{K,V}
    dict::Dict{K,V}
    NoLengthDict{K,V}() where {K,V} = new(Dict{K,V}())
end

Base.iterate(d::NoLengthDict, s...) = iterate(d.dict, s...)
Base.IteratorSize(::Type{<:NoLengthDict}) = Base.SizeUnknown()
Base.eltype(::Type{NoLengthDict{K,V}}) where {K,V} = Pair{K,V}
Base.setindex!(d::NoLengthDict, k, v) = d.dict[k] = v
```

Then in a REPL:
```
julia> x=NoLengthDict{Int,Int}()
NoLengthDict{Int64, Int64}()

julia> x[1]=2
2

julia> x
NoLengthDict{Int64, Int64}(...):Error showing value of type NoLengthDict{Int64, Int64}:
ERROR: MethodError: no method matching length(::NoLengthDict{Int64, Int64})

Closest candidates are:
  length(!Matched::Distributed.WorkerPool)
   @ Distributed /nix/store/xy70mqpgn05sydk1w6ismnc8plbmmhfs-julia-1.10.2/share/julia/stdlib/v1.10/Distributed/src/workerpool.jl:139
  length(!Matched::Core.SimpleVector)
   @ Base essentials.jl:770
  length(!Matched::BitSet)
   @ Base bitset.jl:348
  ...

Stacktrace:
  [1] show(io::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, t::NoLengthDict{Int64, Int64})
    @ Base ./show.jl:170
```

It seems like `Base.show` shouldn't crash in this case.

### PR DESCRIPTION
**Pull request by ararslan**
Also fix interacting with them at the REPL.

Fixes #55931

### CODE REVIEW

File: test/show.jl [hunk 1]
```diff
@@ -2773,3 +2773,19 @@ end
...
+Base.iterate(d::NoLengthDict, s...) = iterate(d.dict, s...)
+Base.IteratorSize(::Type{<:NoLengthDict}) = Base.SizeUnknown()
+Base.eltype(::Type{NoLengthDict{K,V}}) where {K,V} = Pair{K,V}
+Base.setindex!(d::NoLengthDict, v, k) = d.dict[k] = v
+
+# Issue 55931
+@testset "show AbstractDict with unknown length" begin
+    x = NoLengthDict{Int,Int}()
+    x[1] = 2
+    @test sprint((io, x) -> show(io, MIME("text/plain"), x), x) == "NoLengthDict(1 => 2)"
```
Comment (ararslan) [inline test/show.jl:2790]:
```suggestion
    @test endswith(sprint((io, x) -> show(io, MIME("text/plain"), x), x), "NoLengthDict(1 => 2)")
```

File: stdlib/REPL/test/replcompletions.jl [hunk 2]
```diff
@@ -1486,8 +1497,21 @@ test_dict_completion("CompletionFoo.test_customdict")
...
-@test REPLCompletions.dict_identifier_key("test_dict_ℂ[\\", :other) isa Tuple
+struct NoLengthDict{K,V} <: AbstractDict{K,V}
+    dict::Dict{K,V}
+    NoLengthDict{K,V}() where {K,V} = new(Dict{K,V}())
+end
+Base.iterate(d::NoLengthDict, s...) = iterate(d.dict, s...)
+Base.IteratorSize(::Type{<:NoLengthDict}) = Base.SizeUnknown()
+Base.eltype(::Type{NoLengthDict{K,V}}) where {K,V} = Pair{K,V}
+Base.setindex!(d::NoLengthDict, v, k) = d.dict[k] = v
+
```
[the remark this answers is not in the document]
Comment (ararslan) [reply stdlib/REPL/test/replcompletions.jl]: I think you're right, good catch

**Comment by ararslan** [comment]
CI failure is unrelated

**Review by nhz2 — APPROVED** [review]
Looks good.

**Pull request merged**